packages/openapi-to-graphql/src/utils.ts

```diff
@@ -51,6 +51,190 @@ export const mitigations = {
+const MIN_LONG = -9007199254740992
+
+/**
+ * verify that a variable contains a safe int (2^31)
+ */
+export function isSafeInteger(n: unknown): boolean {
+  return (
+    typeof n === 'number' &&
+    isFinite(n) &&
+    Math.floor(n) === n &&
+    n <= MAX_INT &&
+    n >= MIN_INT
+  )
+}
+
+/**
+ * verify that a variable contains a safe long (2^53)
+ */
+
+export function isSafeLong(n: unknown): boolean {
+  return typeof n === 'number' && isFinite(n) && n <= MAX_LONG && n >= MIN_LONG
+}
+
+/**
+ *
+ */
+
+export function isSafeFloat(n: unknown): boolean {
+  return typeof n === 'number' && n % 1 !== 0
+}
+
+/**
+ *
+ */
+
+function toDate(n: string) {
+  const parsed = Date.parse(n)
+  const $ref = new Date()
+
+  $ref.setTime(parsed)
+
+  return (
+    (typeof parsed === 'number' &&
+      parsed !== NaN &&
+      parsed > 0 &&
+      String(parsed).length === 13 &&
+      $ref) ||
+    null
+  )
+}
+
+/**
+ *
+ */
+
+export function serializeDate(n: string) {
+  const date = toDate(n)
+  return date && date.toJSON()
+}
+
+/**
+ * verify that a vriable contains a safe date/date-time string
+ */
+
+export function isSafeDate(n: string): boolean {
+  const date = toDate(n)
+  return date !== null && date.getTime() !== NaN
+}
+
+/**
+ *
+ */
+
+export function isURL(s: string): boolean {
+  let res = null
+  const urlRegex = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z0-9]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g
+  try {
+    res = s.match(urlRegex)
+  } catch (e) {
+    res = null
+  }
+  return res !== null
+}
+
+/**
+ *
+ */
+
+export function isEmail(s: string): boolean {
+  const emailRegex = /^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
+  return emailRegex.test(s)
+}
+
+/**
+ *
+ */
+
+export function isUUIDOrGUID(s: string): boolean {
+  const uuidRegExp = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
+  const guidRegExp = /^(\{){0,1}[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}(\}){0,1}$/gi
+
+  if (s.startsWith('{')) {
+    s = s.substring(1, s.length - 1)
+  }
+
+  return uuidRegExp.test(s) || guidRegExp.test(s)
+}
+
+/**
+ *
+ */
+
+export function ucFirst(s: string): string {
+  if (typeof s !== 'string') {
+    return ''
+  }
+
+  return s.replace(/^./, c => c.toUpperCase())
+}
+
+/**
+ * check if a literal is falsy or not
+ */
+const isLiteralFalsey = (variable): boolean => {
+  return variable === '' || variable === false || variable === 0
+}
+
+/**
+ * provide the name of primitive and/or reference types
+ */
+const checkTypeName = (target, type): boolean => {
+  let typeName = ''
+
+  if (isLiteralFalsey(target)) {
+    typeName = typeof target
+  } else {
+    typeName = '' + (target && target.constructor.name)
+  }
+  return !!(typeName.toLowerCase().indexOf(type) + 1)
+}
+
+/**
+ * get the correct type of a variable
+ */
+export function strictTypeOf(value, type): boolean {
+  let result = false
```

Comment:
This function was created by me (I own the copyright - 😄😀) in 2015. The only references I have are the many codebases I have used it with resounding success. The function has also been tested in multiple instances and ways. It basically uses the variables' constructor name to determine what that type the variable is.

Below are example of its implementation:

```js
const isFunction = (variable) => strictTypeOf(variable, 'function')
const isBlob = (variable) => strictTypeOf(variable, 'blob')
const isSet = (variable) => strictTypeOf(variable, 'set')
const isMap = (variable) => strictTypeOf(variable, 'map')
const isBuffer = (variable) => strictTypeOf(variable, 'buffer')
const isEmitter = (variable) => strictTypeOf(variable, 'eventemitter')
```

Reply:
Also, I modified it to not inspect the _prototype chain_ as such checks are not required for this project ( or are they ? could they be ? ). Here is the original version of the `strictTypeOf` implementation.

```js
/**
 * check if a literal is falsy or not
 */
const isLiteralFalsey = (variable): boolean => {
  return variable === '' || variable === false || variable === 0
}

/**
 * retrieve the name of a types' constructor or up the prototype chain
*/
const getNameByChain = (_type, depth) => {
   if(typeof depth !== 'number'){
      depth = 1;
  }
  let name = _type.name || _type.displayName
  const chainList = [(_type.__proto__ && _type.__proto__.prototype)];
  while(Boolean(chainList[0]) 
          && depth !== 0){
       const variable = chainList[0].constructor;
       if(typeof variable !== 'function'){
           break;
       }
       name = variable.name || variable.displayName
       --depth
       chainList.unshift((variable.__proto__ && variable__proto__.prototype));
  }
  chainList = []
  return name
}

/**
 * provide the name of primitive and/or reference types
 */
const checkTypeName = (target, type): boolean => {
  let typeName = ''
  let match = false
  let depth = 0
  const MAX_DEPTH = 3

  if(typeof type === 'function'){
         type = (getNameByChain(type, depth)).toLowerCase()
  }

  if (isLiteralFalsey(target)) {
    typeName = typeof target
  } else {
    typeName = '' + (target && getNameByChain(target.constructor, depth))
  }

  match = Boolean(typeName.toLowerCase().indexOf(type) + 1)

  while(!match){
      ++depth;
      if(depth === MAX_DEPTH){
           break;
      }
      typeName = '' + (target && getNameByChain(target.constructor, depth))
      match = Boolean(typeName.toLowerCase().indexOf(type) + 1)
  }

  return match;
}

/**
 * get the correct type of a variable
 */
export function strictTypeOf(value, type): boolean {
  let result = false

  type = type || []

  if (typeof type === 'object') {
    if (typeof type.length !== 'number') {
      return result
    }

    let bitPiece = 0

    type = [].slice.call(type)

    type.forEach(_type => {
      localResult = false;
      if (typeof _type === 'function') {
          localResult = value instanceof _type
      }
      bitPiece |= Number(localResult || checkTypeName(value, _type))
    })
    result = Boolean(bitPiece)
  } else {
    if (typeof type === 'function') {
      result = value instanceof type
    }
    result = result || checkTypeName(value, type)
  }
  return result
}
```

Reply:
@isocroft I think it would be helpful to improve the documentation of this function. Both arguments are typed as `any`, but isn't at least the second supposed to be a `string`? The comment for the function states `get the correct type of a variable`. Could you elaborate on the context? It seems we are comparing a runtime value with what is defined in a JSON schema definition here, is that correct? I think it would also help to add some inline comments within this function - at least I am having issues following what it does.

Reply:
@ErikWittern the version of the `strictTypeOf` function pasted above is not the same version used in the repo and in this PR. As it does a lot of things that is not needed for this project (e.g. looking up the prototype chain). Below is the slimmed-down version being used in the repo and in this PR - I have also added comments:

```js

/**
 * check if a literal is falsy or not
 */
const isLiteralFalsey = (variable: unknown): boolean => {
  return variable === '' || variable === false || variable === 0
}

/**
 * check if a variable contains a reference type (not a literal)
 */

const isPrimitive = (arg: unknown): boolean => {
   return typeof arg === 'object' || Boolean(arg) && typeof arg.apply === 'function'
}

/**
 * provide the name of primitive and/or reference types
 */
const checkTypeName = (target: unknown, type: string): boolean => {
  let typeName = ''
  
  // we need to separate checks for literal types and 
  // primitive types so we can speed up performance and
  // keep things simple
  if (isLiteralFalsey(target) || !isPrimitive(target)) {
    // literal 
    typeName = typeof target
  } else {
    // primitive/reference
    typeName = (Object.prototype.toString.call(target)).replace(/^\[object (.+)\]$/, '$1') 
  }

  return Boolean(typeName.toLowerCase().indexOf(type) + 1)
}

/**
 * get the correct type of a variable
 */
export function strictTypeOf(value: unknown, type: string): boolean {

  if (type === 'integer') {
    type = 'number'
  }

  type = type || ''
  return checkTypeName(value, type)
}

```

Reply:
examples:

```js
// swagger/OPenAPI types: 'integer', 'string', 'array', 'object'
strictTypeOf(variable, 'integer')
```

Reply:
@ErikWittern No, it's not correct. The function isn't comparing a runtime value with what is defined in a JSON schema definition. It is checking that the data type of the variable is as intended/expected. I have added comments and references to this function for more clarity.